File: scripts/local-relay-shim.py
#!/usr/bin/env python3
"""
local-relay-shim.py — Python http.server implementation of the Onlook
Phase H/Q manifest endpoint. Used as a drop-in replacement for the Bun
shim because Bun's HTTP layer normalizes well-known headers to
PascalCase (Cache-Control, Content-Type, etc.) which crashes iOS Expo
Go's NSURLSession response handler with EXC_BREAKPOINT (SIGTRAP).

Real `expo start` returns lowercase header names (cache-control,
content-type) and Expo Go loads the manifest fine. Python's
BaseHTTPRequestHandler.send_header(name, value) preserves the exact
case provided — no normalization — so this script can match expo-cli's
wire format byte-for-byte.

PROXY MODE: when EXPO_PROXY_URL is set in env, the manifest endpoint
forwards the upstream expo-cli response 1:1 (status, headers, body).
Used as a debugging bisection — if the phone successfully loads a
manifest proxied from real `expo start`, we know our shim's HTTP layer
is the bug, not the URL or port.

Run via:
    LAN_IP=192.168.0.14 PORT=8787 STORE_DIR=/tmp/cf-builds python3 \\
        scripts/local-relay-shim.py
    LAN_IP=192.168.0.14 PORT=8787 EXPO_PROXY_URL=http://192.168.0.14:8082 \\
        python3 scripts/local-relay-shim.py
"""

from __future__ import annotations

import json
import os
import re
import sys
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse
from urllib.request import Request, urlopen
from urllib.error import URLError

PORT = int(os.environ.get('PORT', '8787'))
STORE_DIR = Path(os.environ.get('STORE_DIR', '/tmp/cf-builds'))
LAN_IP = os.environ.get('LAN_IP', '127.0.0.1')
EXPO_PROXY_URL = os.environ.get('EXPO_PROXY_URL', '')

print(f'[local-relay-shim-py] starting on port {PORT}', flush=True)
print(f'[local-relay-shim-py] store: {STORE_DIR}', flush=True)
print(f'[local-relay-shim-py] LAN URL: http://{LAN_IP}:{PORT}', flush=True)
print(f'[local-relay-shim-py] proxy upstream: {EXPO_PROXY_URL or "(disabled)"}', flush=True)


def bundle_hash_to_uuid_v4(bundle_hash: str) -> str:
    """Deterministic UUID v4 from a 64-char hex sha256 hash."""
    if len(bundle_hash) < 32:
        return bundle_hash
    h = bundle_hash.lower()[:32]
    variant_nibble = ((int(h[16], 16) & 0x3) | 0x8)
    v4 = h[:12] + '4' + h[13:16] + format(variant_nibble, 'x') + h[17:32]
    return f'{v4[0:8]}-{v4[8:12]}-{v4[12:16]}-{v4[16:20]}-{v4[20:32]}'


def build_patched_manifest(bundle_hash: str, fields: dict, built_at: str, platform: str) -> dict:
    """Construct the Expo manifest matching expo-cli's exact field set."""
    debugger_host = f'{LAN_IP}:{PORT}'
    slug = fields['extra']['expoClient']['slug']
    anonymous_id = bundle_hash_to_uuid_v4(bundle_hash)
    scope_key = f'@anonymous/{slug}-{anonymous_id}'

    # launchAsset.url matches expo-cli's exact query string EXCEPT we
    # embed the bundle hash in the entry filename so the relay's bundle
    # route can look it up:
    #   /<hash>.ts.bundle?platform=ios&dev=false&...&transform.bytecode=1
    # Expo Go just fetches whatever URL is here — the entry name doesn't
    # matter to it functionally, only the .bundle suffix.
    bundle_query = (
        f'platform={platform}&dev=false&hot=false&lazy=true&minify=true'
        f'&transform.engine=hermes&transform.bytecode=1&transform.routerRoot=app'
        f'&unstable_transformProfile=hermes-stable'
    )
    launch_asset_url = f'http://{debugger_host}/{bundle_hash}.ts.bundle?{bundle_query}'

    # Normalize createdAt to include milliseconds — strict ISO 8601
    # parsers (and Expo Go's manifest validator) may reject the
    # no-milliseconds form. expo-cli always emits "...Z" with .NNN.
    if isinstance(built_at, str) and built_at.endswith('Z') and '.' not in built_at:
        built_at = built_at[:-1] + '.000Z'

    # Strip fields that expo-cli doesn't include in expoClient
    cleaned_expo_client = {
        k: v for k, v in fields['extra']['expoClient'].items()
        if k not in ('icon', 'runtimeVersion', 'splash')
    }

    return {
        'id': bundle_hash_to_uuid_v4(bundle_hash),
        'createdAt': built_at,
        'runtimeVersion': fields['runtimeVersion'],
        'launchAsset': {
            'key': fields['launchAsset']['key'],
            'contentType': fields['launchAsset']['contentType'],
            'url': launch_asset_url,
        },
        'assets': [],
        'metadata': {},
        'extra': {
            'eas': {},
            'expoClient': {
                **cleaned_expo_client,
                '_internal': {
                    'isDebug': False,
                    'projectRoot': '/private/tmp/onlook-fixture',
                    'dynamicConfigPath': None,
                    'staticConfigPath': '/private/tmp/onlook-fixture/app.json',
                    'packageJsonPath': '/private/tmp/onlook-fixture/package.json',
                },
                'hostUri': debugger_host,
            },
            'expoGo': {
                'debuggerHost': debugger_host,
                'developer': {
                    'tool': 'expo-cli',
                    'projectRoot': '/private/tmp/onlook-fixture',
                },
                'packagerOpts': {'dev': False},
                'mainModuleName': 'index.ts',
            },
            'scopeKey': scope_key,
        },
    }


class RelayHandler(BaseHTTPRequestHandler):
    # Match expo-cli's protocol version (HTTP/1.1 with keep-alive)
    protocol_version = 'HTTP/1.1'
    # Override the default Server header so we don't reveal Python's version.
    server_version = 'expo-cli'
    sys_version = ''

    def log_message(self, format, *args):  # noqa: A002
        # Suppress the default per-request log line; we have our own.
        pass

    def _log_request(self):
        ua = self.headers.get('User-Agent', '?')
        plat = self.headers.get('Expo-Platform', '?')
        rtv = self.headers.get('Expo-Runtime-Version', '?')
        print(
            f'[local-relay-shim-py] {time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())} '
            f'{self.command} {self.path} expo-platform={plat} expo-rtv={rtv} '
            f'ua={ua[:100]}',
            flush=True,
        )
        if self.path.startswith('/manifest/'):
            all_headers = {k.lower(): v for k, v in self.headers.items()}
            print(f'[local-relay-shim-py] ALL headers: {json.dumps(all_headers)}', flush=True)

    def _send_lowercase_response(self, status, headers, body):
        """
        Send a response where header case is preserved EXACTLY as
        provided. send_response() adds Server + Date headers (with
        PascalCase). To prevent that, use send_response_only() which
        ONLY writes the status line, then we add headers manually.
        """
        body_bytes = body if isinstance(body, bytes) else body.encode('utf-8')
        self.send_response_only(status, 'OK')
        for name, value in headers:
            self.send_header(name, value)
        self.send_header('Content-Length', str(len(body_bytes)))
        self.end_headers()
        self.wfile.write(body_bytes)

    def _proxy_to_expo(self):
        """Forward this request to the upstream expo-cli server, return its raw response."""
        if not EXPO_PROXY_URL:
            return False
        upstream = urlparse(EXPO_PROXY_URL)
        upstream_url = f'{EXPO_PROXY_URL}/'
        # Forward request headers — but rewrite Host to upstream's host:port
        fwd_headers = {k: v for k, v in self.headers.items() if k.lower() != 'host'}
        fwd_headers['Host'] = f'{upstream.hostname}:{upstream.port or 80}'
        try:
            req = Request(upstream_url, headers=fwd_headers, method='GET')
            with urlopen(req, timeout=10) as upstream_res:
                # Read raw headers — urllib gives us a list of (name, value)
                # tuples in upstream order with original case preserved.
                raw_headers = upstream_res.headers.items()
                body = upstream_res.read()
                print(
                    f'[local-relay-shim-py] PROXY: forwarded {len(body)} bytes '
                    f'from {upstream_url} (status={upstream_res.status})',
                    flush=True,
                )
                # Re-emit upstream's headers EXACTLY (case preserved)
                # via send_header. Skip Content-Length (we'll re-add).
                filtered = [(k, v) for k, v in raw_headers if k.lower() != 'content-length']
                self._send_lowercase_response(upstream_res.status, filtered, body)
                return True
        except URLError as err:
            print(f'[local-relay-shim-py] PROXY ERROR: {err}', flush=True)
            self._send_lowercase_response(
                502,
                [('Content-Type', 'text/plain')],
                f'relay proxy error: {err}'.encode('utf-8'),
            )
            return True

    def do_GET(self):  # noqa: N802
        self._log_request()

        url = urlparse(self.path)
        path = url.path

        if path == '/health':
            self._send_lowercase_response(
                200,
                [('content-type', 'application/json')],
                json.dumps({'ok': True, 'version': '0.1.0-local-shim-py'}),
            )
            return

        manifest_match = re.match(r'^/manifest/([a-f0-9]{64})$', path)
        if manifest_match:
            bundle_hash = manifest_match.group(1)

            # Proxy mode short-circuits everything
            if EXPO_PROXY_URL and self._proxy_to_expo():
                return

            fields_path = STORE_DIR / bundle_hash / 'manifest-fields.json'
            meta_path = STORE_DIR / bundle_hash / 'meta.json'
            if not fields_path.exists():
                self._send_lowercase_response(
                    404,
                    [('content-type', 'application/json')],
                    json.dumps({'error': f'manifest-fields.json not found for {bundle_hash}'}),
                )
                return

            fields = json.loads(fields_path.read_text())
            built_at = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())
            if meta_path.exists():
                try:
                    meta = json.loads(meta_path.read_text())
                    if isinstance(meta.get('builtAt'), str):
                        built_at = meta['builtAt']
                except json.JSONDecodeError:
                    pass

            # Resolve platform from Expo-Platform header
            header_platform = self.headers.get('Expo-Platform', 'ios')
            platform = 'ios' if header_platform == 'ios' else 'android'

            patched = build_patched_manifest(bundle_hash, fields, built_at, platform)
            manifest_json = json.dumps(patched, separators=(',', ':'))
            boundary = 'formdata-' + bundle_hash[:16]
            body = (
                f'--{boundary}\r\n'
                f'Content-Disposition: form-data; name="manifest"\r\n'
                f'Content-Type: application/json\r\n'
                f'\r\n'
                f'{manifest_json}\r\n'
                f'--{boundary}--\r\n'
            )

            # Headers in expo-cli's exact case + order. Python preserves
            # the case verbatim — no normalization.
            self._send_lowercase_response(
                200,
                [
                    ('expo-protocol-version', '0'),
                    ('expo-sfv-version', '0'),
                    ('cache-control', 'private, max-age=0'),
                    ('content-type', f'multipart/mixed; boundary={boundary}'),
                    ('Connection', 'keep-alive'),
                    ('Keep-Alive', 'timeout=5'),
                ],
                body,
            )
            return

        # Metro-style bundle URL with hash embedded in entry filename:
        # /<hash>.ts.bundle?platform=ios&...
        # Expo Go just sees /<entry>.bundle which is the standard Metro
        # URL pattern; the entry name happens to be the bundle hash.
        bundle_match = re.match(r'^/([a-f0-9]{64})(?:\.ts)?\.bundle$', path)
        if bundle_match:
            qs = parse_qs(url.query)
            query_hash = bundle_match.group(1)
            query_platform = (qs.get('platform') or [''])[0]
            plat = 'android' if query_platform == 'android' else 'ios'
            bundle_js_path = STORE_DIR / query_hash / f'index.{plat}.bundle.js'
            if not bundle_js_path.exists():
                self._send_lowercase_response(
                    404,
                    [('content-type', 'text/plain')],
                    f'relay: no bundle for {query_hash}/{plat}',
                )
                return
            body = bundle_js_path.read_bytes()
            print(
                f'[local-relay-shim-py] served Metro JS bundle '
                f'{query_hash[:12]}/{plat} ({len(body)} bytes)',
                flush=True,
            )
            self._send_lowercase_response(
                200,
                [
                    ('X-Content-Type-Options', 'nosniff'),
                    ('Surrogate-Control', 'no-store'),
                    ('Cache-Control', 'no-store, no-cache, must-revalidate, proxy-revalidate'),
                    ('Pragma', 'no-cache'),
                    ('Expires', '0'),
                    ('Content-Type', 'application/javascript; charset=UTF-8'),
                    ('Connection', 'keep-alive'),
                    ('Keep-Alive', 'timeout=5'),
                ],
                body,
            )
            return

        # No-op endpoints to absorb Expo Go dev-mode side connections
        if path in ('/logs', '/status', '/symbolicate'):
            self._send_lowercase_response(
                200,
                [('content-type', 'application/json')],
                '{}',
            )
            return

        self._send_lowercase_response(
            404,
            [('content-type', 'text/plain')],
            'not found',
        )

    def do_POST(self):  # noqa: N802
        self._log_request()
        # Absorb POSTs to /logs and /symbolicate from Expo Go dev mode
        if self.path in ('/logs', '/symbolicate'):
            content_length = int(self.headers.get('Content-Length', '0'))
            if content_length:
                self.rfile.read(content_length)
            self._send_lowercase_response(
                200,
                [('content-type', 'application/json')],
                '{}',
            )
            return
        self._send_lowercase_response(
            404,
            [('content-type', 'text/plain')],
            'not found',
        )

    def do_OPTIONS(self):  # noqa: N802
        self._log_request()
        self._send_lowercase_response(
            200,
            [
                ('Access-Control-Allow-Origin', '*'),
                ('Access-Control-Allow-Methods', 'GET, POST, OPTIONS'),
                ('Access-Control-Allow-Headers', '*'),
            ],
            b'',
        )


def main():
    server = ThreadingHTTPServer(('0.0.0.0', PORT), RelayHandler)
    print(f'[local-relay-shim-py] listening on 0.0.0.0:{PORT}', flush=True)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('[local-relay-shim-py] shutting down', flush=True)
        server.server_close()


if __name__ == '__main__':
    main()

File: scripts/local-relay-shim.ts
#!/usr/bin/env bun
/**
 * local-relay-shim.ts — local replacement for cf-expo-relay Worker.
 *
 * Reads the bundle artifact set from the local-builder-shim's store
 * (/tmp/cf-builds/<hash>/manifest-fields.json + meta.json) and constructs
 * a real Expo Updates v2 manifest using the SAME buildManifest() function
 * the cf-expo-relay Worker uses (apps/cf-expo-relay/src/manifest-builder.ts).
 *
 * Endpoints (mirrors plans/expo-browser-relay-manifest.md):
 *   GET /health                     → { ok, version }
 *   GET /manifest/:bundleHash       → Expo Updates v2 manifest JSON
 *
 * The manifest's launchAsset.url points at the local-builder-shim
 * (CF_ESM_CACHE_URL env var, defaults to http://${LAN_IP}:8788).
 *
 * Run via:
 *   bun scripts/local-relay-shim.ts
 *   PORT=8787 LAN_IP=192.168.0.14 CF_ESM_CACHE_URL=http://192.168.0.14:8788 \
 *     bun scripts/local-relay-shim.ts
 */

import { readFile } from 'fs/promises';
import { existsSync } from 'fs';
import { join } from 'path';
import { buildManifest, type ManifestFields } from '../apps/cf-expo-relay/src/manifest-builder';

const PORT = parseInt(process.env.PORT ?? '8787', 10);
const STORE_DIR = process.env.STORE_DIR ?? '/tmp/cf-builds';
const LAN_IP = process.env.LAN_IP ?? '127.0.0.1';
const CF_ESM_CACHE_URL = process.env.CF_ESM_CACHE_URL ?? `http://${LAN_IP}:8788`;
const VERSION = '0.1.0-local-shim';

/**
 * Convert a 64-char hex SHA256 hash into a deterministic UUID v4 string.
 * Used to derive a stable scopeKey suffix for the manifest's
 * `extra.scopeKey` field. Mirrors the helper in cf-expo-relay's
 * manifest-builder.ts.
 */
function bundleHashToUuidV4Local(bundleHash: string): string {
    if (bundleHash.length < 32) return bundleHash;
    const h = bundleHash.toLowerCase().slice(0, 32);
    const v4 =
        h.slice(0, 12) +
        '4' +
        h.slice(13, 16) +
        (((parseInt(h[16] ?? '0', 16) & 0x3) | 0x8).toString(16)) +
        h.slice(17, 32);
    return `${v4.slice(0, 8)}-${v4.slice(8, 12)}-${v4.slice(12, 16)}-${v4.slice(16, 20)}-${v4.slice(20, 32)}`;
}

console.log(`[local-relay-shim] starting on port ${PORT}`);
console.log(`[local-relay-shim] store: ${STORE_DIR}`);
console.log(`[local-relay-shim] cache URL: ${CF_ESM_CACHE_URL}`);
console.log(`[local-relay-shim] LAN URL: http://${LAN_IP}:${PORT}`);

Bun.serve({
    port: PORT,
    hostname: '0.0.0.0',
    async fetch(req: Request) {
        const url = new URL(req.url);
        const path = url.pathname;
        const method = req.method;

        // Log every request so we can prove the phone reached us, and what
        // it asked for. Expo Go SDK 50+ sends User-Agent + Expo-Platform.
        const ua = req.headers.get('user-agent') ?? '?';
        const expoPlatform = req.headers.get('expo-platform') ?? '?';
        const expoSdk = req.headers.get('expo-sdk-version') ?? '?';
        const expoRtv = req.headers.get('expo-runtime-version') ?? '?';
        const remote = req.headers.get('x-forwarded-for') ?? '(local)';
        console.log(
            `[local-relay-shim] ${new Date().toISOString()} ${method} ${path} ` +
            `expo-platform=${expoPlatform} expo-sdk=${expoSdk} expo-rtv=${expoRtv} ` +
            `from=${remote} ua=${ua.slice(0, 100)}`,
        );

        // Dump ALL request headers for /manifest/* requests so we can spot
        // code-signing, channel, expect-signature, etc. that Expo Go SDK 54
        // sends but we don't normally surface.
        if (path.startsWith('/manifest/')) {
            const allHeaders: Record<string, string> = {};
            req.headers.forEach((value, key) => {
                allHeaders[key] = value;
            });
            console.log(`[local-relay-shim] ALL headers: ${JSON.stringify(allHeaders)}`);
        }

        // CORS — Expo Go iOS client preflights manifest fetches
        const cors = {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, OPTIONS',
            'Access-Control-Allow-Headers': '*',
        };
        if (method === 'OPTIONS') return new Response(null, { headers: cors });

        try {
            if (path === '/health' && method === 'GET') {
                return Response.json({ ok: true, version: VERSION }, { headers: cors });
            }

            // Match /manifest/<hash> with optional query params so callers
            // can cache-bust iOS NSURLSession by appending ?v=<anything>.
            const manifestMatch = path.match(/^\/manifest\/([a-f0-9]{64})$/);
            if (manifestMatch && method === 'GET') {
                const bundleHash = manifestMatch[1];
                const fieldsPath = join(STORE_DIR, bundleHash, 'manifest-fields.json');
                const metaPath = join(STORE_DIR, bundleHash, 'meta.json');

                if (!existsSync(fieldsPath)) {
                    return Response.json(
                        { error: `manifest-fields.json not found for bundleHash ${bundleHash}` },
                        { status: 404, headers: cors },
                    );
                }

                const fieldsRaw = await readFile(fieldsPath, 'utf8');
                let fields: ManifestFields;
                try {
                    fields = JSON.parse(fieldsRaw) as ManifestFields;
                } catch {
                    return Response.json({ error: 'malformed manifest-fields.json' }, { status: 502, headers: cors });
                }

                let builtAt = new Date().toISOString();
                if (existsSync(metaPath)) {
                    try {
                        const meta = JSON.parse(await readFile(metaPath, 'utf8'));
                        if (typeof meta?.builtAt === 'string') builtAt = meta.builtAt;
                    } catch {
                        // best-effort — fall back to now
                    }
                }

                // Resolve target platform from the Expo Go header (or
                // ?platform= for curl-friendly testing). Mirrors the
                // canonical resolvePlatform() in apps/cf-expo-relay.
                const headerPlatform = req.headers.get('expo-platform');
                let platform: 'ios' | 'android' = 'android';
                if (headerPlatform === 'ios') platform = 'ios';
                else if (headerPlatform === 'android') platform = 'android';
                else {
                    const queryPlatform = url.searchParams.get('platform');
                    if (queryPlatform === 'ios') platform = 'ios';
                    else if (queryPlatform === 'android') platform = 'android';
                }

                const manifest = buildManifest({
                    bundleHash,
                    cfEsmCacheUrl: CF_ESM_CACHE_URL,
                    fields,
                    builtAt,
                    platform,
                });

                // Patch the manifest to match real `expo start`'s shape
                // BYTE-FOR-BYTE. Verified against the actual response from
                // `npx expo start --lan` running locally on this Mac on
                // 2026-04-09 — see /tmp/expo-cli-manifest.txt for the
                // captured baseline.
                const debuggerHost = `${LAN_IP}:${PORT}`;
                const slug = manifest.extra.expoClient.slug;

                // Build the launchAsset URL in Metro's exact convention so
                // Expo Go's URL parser doesn't assert on an unfamiliar
                // shape: /<mainModule>.bundle?platform=...&dev=false&...
                // Includes &hash=<bundleHash> so the relay's bundle route
                // can look up the right artifact dir.
                const bundleQuery = new URLSearchParams({
                    platform: 'ios',
                    dev: 'false',
                    hot: 'false',
                    lazy: 'true',
                    minify: 'true',
                    'transform.engine': 'hermes',
                    'transform.bytecode': '0',
                    'transform.routerRoot': 'app',
                    unstable_transformProfile: 'hermes-stable',
                    hash: bundleHash,
                });
                const launchAssetUrl = `http://${debuggerHost}/index.ts.bundle?${bundleQuery.toString()}`;
                // Real expo-cli generates a per-anonymous-id scopeKey of
                // the form `@anonymous/<slug>-<uuid>`. We generate ours
                // deterministically from the bundle hash so the same
                // build always produces the same scopeKey.
                const anonymousId = bundleHashToUuidV4Local(bundleHash);
                const scopeKey = `@anonymous/${slug}-${anonymousId}`;

                const patchedManifest = {
                    ...manifest,
                    launchAsset: {
                        ...manifest.launchAsset,
                        url: launchAssetUrl,
                    },
                    extra: {
                        eas: {},
                        expoClient: {
                            ...manifest.extra.expoClient,
                            // _internal is REQUIRED — expo-cli always sets it
                            // and Expo Go's manifest parser may assert on it.
                            _internal: {
                                isDebug: false,
                                projectRoot: '/private/tmp/onlook-fixture',
                                dynamicConfigPath: null,
                                staticConfigPath: '/private/tmp/onlook-fixture/app.json',
                                packageJsonPath: '/private/tmp/onlook-fixture/package.json',
                            },
                            hostUri: debuggerHost,
                        },
                        expoGo: {
                            debuggerHost,
                            developer: {
                                tool: 'expo-cli',
                                projectRoot: '/private/tmp/onlook-fixture',
                            },
                            packagerOpts: { dev: false },
                            // mainModuleName INCLUDES the .ts extension —
                            // expo-cli sends `index.ts`, NOT `index`. This
                            // is one of the byte-diffs against our previous
                            // attempt.
                            mainModuleName: 'index.ts',
                        },
                        scopeKey,
                    },
                };

                // Match `expo start`'s exact multipart format. Critical
                // differences from the previous version of this shim:
                //   1. NO CORS headers — expo-cli doesn't send any
                //      Access-Control-Allow-* headers and iOS NSURLSession's
                //      response handler may assert on their presence in a
                //      dev-mode manifest response.
                //   2. Boundary format `formdata-<16 hex>` (not
                //      OnlookExpoManifestBoundary...)
                //   3. Connection: keep-alive + Keep-Alive: timeout=5
                //   4. Lowercase header names where expo-cli uses lowercase
                //      (content-type, cache-control, expo-*).
                //   5. Per-part: Content-Disposition + Content-Type (both)
                const boundary = 'formdata-' + bundleHash.slice(0, 16);
                const manifestJson = JSON.stringify(patchedManifest);
                const body =
                    `--${boundary}\r\n` +
                    `Content-Disposition: form-data; name="manifest"\r\n` +
                    `Content-Type: application/json\r\n` +
                    `\r\n` +
                    `${manifestJson}\r\n` +
                    `--${boundary}--\r\n`;

                return new Response(body, {
                    headers: {
                        // Headers in the EXACT order + case expo-cli uses,
                        // no CORS, with Connection: keep-alive.
                        'expo-protocol-version': '0',
                        'expo-sfv-version': '0',
                        'cache-control': 'private, max-age=0',
                        'content-type': `multipart/mixed; boundary=${boundary}`,
                        'Connection': 'keep-alive',
                        'Keep-Alive': 'timeout=5',
                    },
                });
            }

            // Metro-style bundle URL: /<entry>.bundle?platform=...&hash=...
            // Expo Go fetches this from the manifest's launchAsset.url.
            // We serve the pre-Hermes Metro JS output (.bundle.js) NOT
            // the Hermes bytecode — real `expo start` does the same thing
            // even when the URL has transform.bytecode=1, because Expo
            // Go's iOS dev runtime evaluates JS via JSC, not Hermes.
            const metroBundleMatch = path.match(/^\/[^\/]+\.bundle$/);
            if (metroBundleMatch && method === 'GET') {
                const queryHash = url.searchParams.get('hash');
                const queryPlatform = url.searchParams.get('platform');
                if (!queryHash || !/^[a-f0-9]{64}$/.test(queryHash)) {
                    return new Response('relay: missing or invalid &hash=<sha256>', {
                        status: 400,
                    });
                }
                const plat: 'ios' | 'android' =
                    queryPlatform === 'android' ? 'android' : 'ios';
                const bundleJsPath = join(
                    STORE_DIR,
                    queryHash,
                    `index.${plat}.bundle.js`,
                );
                if (!existsSync(bundleJsPath)) {
                    return new Response(`relay: no bundle for hash=${queryHash} plat=${plat}`, {
                        status: 404,
                    });
                }
                const body = await readFile(bundleJsPath);
                console.log(
                    `[local-relay-shim] served Metro JS bundle ${queryHash.slice(0, 12)}/${plat} (${body.length} bytes)`,
                );
                return new Response(body, {
                    headers: {
                        // Headers match real expo-cli's bundle response
                        'X-Content-Type-Options': 'nosniff',
                        'Surrogate-Control': 'no-store',
                        'Cache-Control': 'no-store, no-cache, must-revalidate, proxy-revalidate',
                        'Pragma': 'no-cache',
                        'Expires': '0',
                        'Content-Type': 'application/javascript; charset=UTF-8',
                        'Connection': 'keep-alive',
                        'Keep-Alive': 'timeout=5',
                    },
                });
            }

            // No-op endpoints Expo Go's dev-mode side connections might
            // hit. Return 200 with an empty body so the iOS HTTP client
            // doesn't retry / hang waiting for one of them. None of these
            // are real Metro features — Onlook just needs the responses
            // to NOT be 404s so Expo Go's URL session delegate doesn't
            // assert in its retry loop.
            if (
                method === 'GET' &&
                (path === '/logs' || path === '/status' || path === '/symbolicate')
            ) {
                return new Response('{}', {
                    headers: { ...cors, 'Content-Type': 'application/json' },
                });
            }
            if (method === 'POST' && (path === '/logs' || path === '/symbolicate')) {
                return new Response('{}', {
                    headers: { ...cors, 'Content-Type': 'application/json' },
                });
            }

            return new Response('not found', { status: 404, headers: cors });
        } catch (err) {
            console.error('[local-relay-shim]', err);
            return Response.json({ error: String(err) }, { status: 500, headers: cors });
        }
    },
});
